File: android/app/src/main/java/com/AppPlatformReactNativeApp/appplatformble/AppPlatformBLEModule.kt
package com.AppPlatformReactNativeApp.trakplusble
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.modules.core.DeviceEventManagerModule
import kotlin.math.max
class AppDeviceBLEModule(
  reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener, AppDeviceBLECore.Delegate {
  private val core = AppDeviceBLECore(reactContext.applicationContext)
  private val eventBufferLock = Any()
  private val eventBuffer: ArrayDeque<BufferedEvent> = ArrayDeque()
  private val maxBufferSize = 100
  private val pendingOverflows: MutableList<Map<String, Any?>> = mutableListOf()
  private val maxPendingOverflows = 10
  @Volatile
  private var listenerCount = 0
  @Volatile
  private var totalBuffered = 0
  @Volatile
  private var totalDropped = 0
  @Volatile
  private var lastConnectionStateEvent: Map<String, Any?>? = null
  @Volatile
  private var hasReplayedConnectionStateForCurrentObserverCycle = false
  init {
    reactContext.addLifecycleEventListener(this)
    core.initialize(this)
  }
  override fun getName(): String = "AppDeviceBLE"
  @ReactMethod
  fun addListener(eventName: String) {
    listenerCount += 1
    flushBufferedEvents()
    if (
      eventName == EVENT_CONNECTION_STATE &&
      !hasReplayedConnectionStateForCurrentObserverCycle
    ) {
      lastConnectionStateEvent?.let { sendEventOrBuffer(EVENT_CONNECTION_STATE, it) }
      hasReplayedConnectionStateForCurrentObserverCycle = true
    }
  }
  @ReactMethod
  fun removeListeners(count: Int) {
    listenerCount = max(0, listenerCount - count)
    if (listenerCount == 0) {
      hasReplayedConnectionStateForCurrentObserverCycle = false
    }
  }
  @ReactMethod
  fun startScan(broadScan: Boolean) {
    core.startScan(broadScan)
  }
  @ReactMethod
  fun stopScan() {
    core.stopScan()
  }
  @ReactMethod
  fun connect(deviceId: String) {
    core.connect(deviceId)
  }
  @ReactMethod
  fun disconnect() {
    core.disconnect()
  }
  @ReactMethod
  fun write(base64Data: String, promise: Promise) {
    core.write(base64Data) { accepted, errorCode, errorMessage ->
      if (accepted) {
        promise.resolve(null)
      } else {
        promise.reject(errorCode ?: "WRITE_REJECTED", errorMessage ?: "Write rejected")
      }
    }
  }
  @ReactMethod
  fun getConnectionState(promise: Promise) {
    core.getConnectionState { snapshot ->
      promise.resolve(snapshot.toWritableMap())
    }
  }
  @ReactMethod
  fun checkSystemConnections() {
    core.checkSystemConnections()
  }
  @ReactMethod
  fun setKnownPeripheralIds(ids: ReadableArray) {
    val list = mutableListOf<String>()
    for (index in 0 until ids.size()) {
      if (!ids.isNull(index)) {
        list.add(ids.getString(index))
      }
    }
    core.setKnownPeripheralIds(list)
  }
  @ReactMethod
  fun setDeviceSleepFlag() {
    core.setDeviceSleepFlag()
  }
  @ReactMethod
  fun getBufferDiagnostics(promise: Promise) {
    val diagnostics = synchronized(eventBufferLock) {
      mapOf(
        "currentBufferSize" to eventBuffer.size,
        "totalBuffered" to totalBuffered,
        "totalDropped" to totalDropped,
        "hasListeners" to hasListeners(),
        "pendingOverflowCount" to pendingOverflows.size,
      )
    }
    promise.resolve(diagnostics.toWritableMap())
  }
  override fun onHostResume() {
    core.checkSystemConnections()
  }
  override fun onHostPause() {
  }
  override fun onHostDestroy() {
    core.destroy()
  }
  override fun invalidate() {
    super.invalidate()
    reactApplicationContext.removeLifecycleEventListener(this)
    core.destroy()
  }
  override fun onConnectionStateChanged(event: ConnectionStateEvent) {
    val payload = mapOf(
      "state" to event.state.wireValue,
      "deviceId" to event.deviceId,
      "reason" to event.reason?.wireValue,
      "timestamp" to event.timestampMs.toDouble(),
      "foregroundServiceActive" to event.foregroundServiceActive,
    ).filterValues { it != null }
    lastConnectionStateEvent = payload
    sendEventOrBuffer(EVENT_CONNECTION_STATE, payload)
  }
  override fun onDataReceived(event: DataReceivedEvent) {
    val payload = mapOf(
      "data" to event.dataBase64,
      "deviceId" to event.deviceId,
      "timestamp" to event.timestampMs.toDouble(),
    )
    sendEventOrBuffer(EVENT_DATA_RECEIVED, payload)
  }
  override fun onBondingLost(deviceId: String) {
    sendEventOrBuffer(EVENT_BONDING_LOST, mapOf("deviceId" to deviceId))
  }
  override fun onBluetoothStateChanged(state: BluetoothAdapterState) {
    sendEventOrBuffer(EVENT_BLUETOOTH_STATE, mapOf("state" to state.wireValue))
  }
  override fun onDeviceDiscovered(device: DiscoveredDevice) {
    val payload = mapOf(
      "id" to device.id,
      "name" to device.name,
      "rssi" to device.rssi,
      "isConnectable" to device.isConnectable,
    )
    sendEventOrBuffer(EVENT_DEVICE_FOUND, payload)
  }
  override fun onOperationRejected(event: OperationRejectedEvent) {
    val payload = mapOf(
      "operation" to event.operation.wireValue,
      "reason" to event.reason.wireValue,
      "detail" to event.detail,
      "deviceId" to event.deviceId,
      "timestamp" to event.timestampMs.toDouble(),
    ).filterValues { it != null }
    sendEventOrBuffer(EVENT_OPERATION_REJECTED, payload)
  }
  private fun hasListeners(): Boolean = listenerCount > 0
  private fun canEmitToJs(): Boolean {
    return hasListeners() && reactApplicationContext.hasActiveReactInstance()
  }
  private fun sendEventOrBuffer(name: String, body: Map<String, Any?>) {
    if (canEmitToJs()) {
      emitEvent(name, body)
      return
    }
    val overflowSnapshot = synchronized(eventBufferLock) {
      val dropped = if (eventBuffer.size >= maxBufferSize) eventBuffer.removeFirst() else null
      eventBuffer.addLast(BufferedEvent(name, body, System.currentTimeMillis()))
      totalBuffered += 1
      if (dropped != null) {
        totalDropped += 1
      }
      dropped
    }
    if (overflowSnapshot != null) {
      val overflowPayload = mapOf(
        "droppedEventName" to overflowSnapshot.name,
        "droppedEventTimestamp" to overflowSnapshot.timestampMs.toDouble(),
        "totalDropped" to totalDropped,
        "bufferSize" to synchronized(eventBufferLock) { eventBuffer.size },
        "timestamp" to System.currentTimeMillis().toDouble(),
      )
      if (canEmitToJs()) {
        emitEvent(EVENT_BUFFER_OVERFLOW, overflowPayload)
      } else {
        synchronized(eventBufferLock) {
          pendingOverflows.add(overflowPayload)
          if (pendingOverflows.size > maxPendingOverflows) {
            pendingOverflows.removeAt(0)
          }
        }
      }
    }
  }
  private fun flushBufferedEvents() {
    if (!canEmitToJs()) {
      return
    }
    val events: List<BufferedEvent>
    val overflows: List<Map<String, Any?>>
    synchronized(eventBufferLock) {
      if (eventBuffer.isEmpty() && pendingOverflows.isEmpty()) {
        return
      }
      events = eventBuffer.toList()
      eventBuffer.clear()
      overflows = pendingOverflows.toList()
      pendingOverflows.clear()
    }
    events.forEach { buffered ->
      emitEvent(buffered.name, buffered.body)
    }
    overflows.forEach { payload ->
      emitEvent(EVENT_BUFFER_OVERFLOW, payload)
    }
  }
  private fun emitEvent(name: String, body: Map<String, Any?>) {
    val context = reactApplicationContext
    if (!context.hasActiveReactInstance()) {
      return
    }
    context.runOnUiQueueThread {
      if (!context.hasActiveReactInstance()) {
        return@runOnUiQueueThread
      }
      context
        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
        .emit(name, body.toWritableMap())
    }
  }
  private fun Map<String, Any?>.toWritableMap() = Arguments.makeNativeMap(this)
  private data class BufferedEvent(
    val name: String,
    val body: Map<String, Any?>,
    val timestampMs: Long,
  )
  companion object {
    private const val EVENT_CONNECTION_STATE = "onConnectionStateChange"
    private const val EVENT_DATA_RECEIVED = "onDataReceived"
    private const val EVENT_BONDING_LOST = "onBondingLost"
    private const val EVENT_BLUETOOTH_STATE = "onBluetoothStateChange"
    private const val EVENT_DEVICE_FOUND = "onDeviceFound"
    private const val EVENT_BUFFER_OVERFLOW = "onBufferOverflow"
    private const val EVENT_OPERATION_REJECTED = "onOperationRejected"
  }
}

File: ios/AppPlatform/AppPlatformBLE/AppPlatformBLEModule.swift
import Foundation
import React
/
struct DroppedEventInfo {
    let eventName: String
    let timestamp: Date
}
/
/
/
/
private final class EventBuffer {
    /
    struct BufferedEvent {
        let name: String
        let body: [String: Any]?
        let timestamp: Date
    }
    /
    private let maxSize: Int
    /
    private var buffer: [BufferedEvent] = []
    /
    private let lock = NSLock()
    /
    private(set) var totalBuffered: Int = 0
    private(set) var totalDropped: Int = 0
    /
    var onOverflow: ((DroppedEventInfo) -> Void)?
    init(maxSize: Int = 100) {
        self.maxSize = maxSize
    }
    /
    /
    /
    /
    func add(name: String, body: [String: Any]?) {
        lock.lock()
        var droppedEvent: BufferedEvent? = nil
        if buffer.count >= maxSize {
            droppedEvent = buffer.removeFirst()
            totalDropped += 1
            print("[EventBuffer] Buffer full, dropped oldest event '\(droppedEvent?.name ?? "unknown")'. Total dropped: \(totalDropped)")
        }
        buffer.append(BufferedEvent(name: name, body: body, timestamp: Date()))
        totalBuffered += 1
        lock.unlock()
        if let dropped = droppedEvent, let callback = onOverflow {
            callback(DroppedEventInfo(eventName: dropped.name, timestamp: dropped.timestamp))
        }
    }
    /
    /
    func flush() -> [BufferedEvent] {
        lock.lock()
        defer { lock.unlock() }
        let events = buffer
        buffer.removeAll()
        return events
    }
    /
    var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return buffer.count
    }
}
/
/
/
/
/
/
/
/
/
/
/
/
/
/
/
/
@objc(AppDeviceBLE)
class AppDeviceBLEModule: RCTEventEmitter {
    /
    private let core = AppDeviceBLECore.shared
    /
    private var hasListeners = false
    /
    /
    private let eventBuffer = EventBuffer(maxSize: 100)
    /
    private var lastConnectionState: ConnectionStateEvent?
    private var pendingOverflows: [[String: Any]] = []
    private let overflowLock = NSLock()
    private let maxPendingOverflows = 10
    override init() {
        super.init()
        core.delegate = self
        eventBuffer.onOverflow = { [weak self] droppedInfo in
            guard let self = self else { return }
            self.log("BUFFER OVERFLOW: Dropped event '\(droppedInfo.eventName)' from \(droppedInfo.timestamp)")
            let overflowPayload: [String: Any] = [
                "droppedEventName": droppedInfo.eventName,
                "droppedEventTimestamp": droppedInfo.timestamp.timeIntervalSince1970 * 1000,
                "totalDropped": self.eventBuffer.totalDropped,
                "bufferSize": self.eventBuffer.count,
                "timestamp": Date().timeIntervalSince1970 * 1000
            ]
            if self.hasListeners {
                self.sendEvent(withName: "onBufferOverflow", body: overflowPayload)
            } else {
                self.overflowLock.lock()
                self.pendingOverflows.append(overflowPayload)
                if self.pendingOverflows.count > self.maxPendingOverflows {
                    self.pendingOverflows.removeFirst()
                }
                self.overflowLock.unlock()
            }
        }
        log("AppDeviceBLEModule initialized with event buffering enabled")
    }
    /
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    /
    override func supportedEvents() -> [String]! {
        return [
            "onConnectionStateChange",
            "onDataReceived",
            "onBondingLost",
            "onBluetoothStateChange",
            "onDeviceFound",
            "onBufferOverflow",
            "onOperationRejected"
        ]
    }
    /
    /
    override func startObserving() {
        hasListeners = true
        log("JS started observing events")
        flushBufferedEvents()
        if let lastState = lastConnectionState {
            log("Sending cached connection state to new listener: \(lastState.state.rawValue)")
            sendEvent(withName: "onConnectionStateChange", body: lastState.toDictionary())
        }
    }
    /
    override func stopObserving() {
        hasListeners = false
        log("JS stopped observing events")
    }
    /
    /
    /
    /
    /
    /
    private func flushBufferedEvents() {
        let bufferedEvents = eventBuffer.flush()
        overflowLock.lock()
        let overflows = pendingOverflows
        pendingOverflows.removeAll()
        overflowLock.unlock()
        if bufferedEvents.isEmpty && overflows.isEmpty {
            log("No buffered events or overflow notifications to flush")
            return
        }
        log("Flushing \(bufferedEvents.count) buffered events and \(overflows.count) overflow notifications to JS")
        for event in bufferedEvents {
            let ageMs = Int(Date().timeIntervalSince(event.timestamp) * 1000)
            log("Flushing buffered event '\(event.name)' (age: \(ageMs)ms)")
            sendEvent(withName: event.name, body: event.body)
        }
        for payload in overflows {
            log("Flushing deferred overflow notification")
            sendEvent(withName: "onBufferOverflow", body: payload)
        }
        log("Buffer flush complete. Stats: totalBuffered=\(eventBuffer.totalBuffered), totalDropped=\(eventBuffer.totalDropped)")
    }
    /
    /
    private func sendEventOrBuffer(name: String, body: [String: Any]?) {
        if hasListeners {
            sendEvent(withName: name, body: body)
        } else {
            log("Buffering event '\(name)' (no JS listeners yet)")
            eventBuffer.add(name: name, body: body)
        }
    }
    /
    /
    @objc
    func startScan(_ broadScan: Bool) {
        log("startScan called from JS (broadScan=\(broadScan))")
        DispatchQueue.main.async {
            self.core.startScan(broadScan: broadScan)
        }
    }
    /
    @objc
    func stopScan() {
        log("stopScan called from JS")
        DispatchQueue.main.async {
            self.core.stopScan()
        }
    }
    /
    /
    /
    @objc
    func connect(_ uuid: String) {
        log("connect called from JS: \(uuid)")
        DispatchQueue.main.async {
            self.core.connect(uuid)
        }
    }
    /
    @objc
    func disconnect() {
        log("disconnect called from JS")
        DispatchQueue.main.async {
            self.core.disconnect()
        }
    }
    /
    /
    /
    /
    /
    /
    /
    /
    /
    /
    @objc
    func write(_ base64Data: String,
               resolver resolve: @escaping RCTPromiseResolveBlock,
               rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let data = Data(base64Encoded: base64Data) else {
            log("write: Invalid base64 data")
            sendEventOrBuffer(name: "onOperationRejected", body: OperationRejectedEvent(
                operation: .write,
                reason: .invalidPayload,
                detail: "Invalid base64 payload",
                deviceId: core.connectedPeripheral?.identifier.uuidString
            ).toDictionary())
            reject("INVALID_PAYLOAD", "Invalid base64 payload", nil)
            return
        }
        DispatchQueue.main.async {
            let accepted = self.core.write(data)
            if accepted {
                resolve(nil)
            } else {
                reject("WRITE_REJECTED", "Write rejected - not in READY state", nil)
            }
        }
    }
    /
    /
    /
    /
    /
    @objc
    func getConnectionState(_ resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            let state = self.core.getConnectionState()
            resolve(state)
        }
    }
    /
    /
    @objc
    func checkSystemConnections() {
        log("checkSystemConnections called from JS")
        DispatchQueue.main.async {
            self.core.checkSystemConnections()
        }
    }
    /
    /
    /
    @objc
    func setKnownPeripheralIds(_ ids: [String]) {
        log("setKnownPeripheralIds called from JS: \(ids.count) IDs")
        DispatchQueue.main.async {
            self.core.setKnownPeripheralIds(ids)
        }
    }
    /
    @objc
    func setDeviceSleepFlag() {
        log("setDeviceSleepFlag called from JS")
        DispatchQueue.main.async {
            self.core.setDeviceSleepFlag()
        }
    }
    /
    @objc
    func getBufferDiagnostics(_ resolve: @escaping RCTPromiseResolveBlock,
                              rejecter reject: @escaping RCTPromiseRejectBlock) {
        overflowLock.lock()
        let pendingCount = pendingOverflows.count
        overflowLock.unlock()
        resolve([
            "currentBufferSize": eventBuffer.count,
            "totalBuffered": eventBuffer.totalBuffered,
            "totalDropped": eventBuffer.totalDropped,
            "hasListeners": hasListeners,
            "pendingOverflowCount": pendingCount
        ])
    }
    private func log(_ message: String) {
        print("[AppDeviceBLEModule] \(message)")
    }
}
extension AppDeviceBLEModule: AppDeviceBLECoreDelegate {
    func connectionStateChanged(_ event: ConnectionStateEvent) {
        log("Connection state changed: \(event.state.rawValue)")
        lastConnectionState = event
        sendEventOrBuffer(name: "onConnectionStateChange", body: event.toDictionary())
    }
    func dataReceived(_ event: DataReceivedEvent) {
        sendEventOrBuffer(name: "onDataReceived", body: event.toDictionary())
    }
    func bondingLost(deviceId: String) {
        log("Bonding lost for device: \(deviceId)")
        sendEventOrBuffer(name: "onBondingLost", body: ["deviceId": deviceId])
    }
    func bluetoothStateChanged(_ state: BluetoothAdapterState) {
        log("Bluetooth state changed: \(state.rawValue)")
        sendEventOrBuffer(name: "onBluetoothStateChange", body: ["state": state.rawValue])
    }
    func deviceDiscovered(_ device: DiscoveredDevice) {
        log("Device discovered: \(device.id)")
        sendEventOrBuffer(name: "onDeviceFound", body: [
            "id": device.id,
            "name": device.name ?? NSNull(),
            "rssi": device.rssi,
            "isConnectable": device.isConnectable
        ])
    }
    func operationRejected(_ event: OperationRejectedEvent) {
        log("Operation rejected: \(event.operation.rawValue) (\(event.reason.rawValue))")
        sendEventOrBuffer(name: "onOperationRejected", body: event.toDictionary())
    }
}
